src/glum/_cd_fast.pyx

```diff
@@ -204,6 +205,139 @@ def enet_coordinate_descent_gram(int[::1] active_set,
 
     return np.asarray(w), norm_min_subgrad, max_min_subgrad, tol, n_iter + 1
 
+def enet_coordinate_descent_gram_diag_fisher(
+                                 X,
+                                 np.ndarray[np.float64_t, ndim=1] hessian_rows,
+                                 int[::1] active_set,
+                                 floating[::1] w,
+                                 floating[::1] P1,
+                                 floating[::1] q,
+                                 int max_iter, floating tol, object rng,
+                                 bint intercept, bint random,
+                                 bint has_lower_bounds,
+                                 floating[:] lower_bounds,
+                                 bint has_upper_bounds,
+                                 floating[:] upper_bounds):
+    """Cython version of the coordinate descent algorithm
+        for Elastic-Net regression
+        We minimize
+        (1/2) * w^T Q w - q^T w + P1 norm(w, 1)
+        which amounts to the L1 problem when:
+        Q = X^T X (Gram matrix)
+        q = X^T y
+
+        This version, for diag_fisher=True, never calculates the entire Hessian matrix; 
+        only rows, when necessary. The inefficiency here is that we must re-calculate 
+        the rows of Q for every iteration up to max_iter.
+    """
+
+    cdef unsigned int n_active_features = active_set.shape[0]
+
+    cdef floating w_ii
+    cdef floating P1_ii
+    cdef floating qii_temp
+    cdef floating d_w_max
+    cdef floating w_max
+    cdef floating d_w_ii
+    cdef floating d_w_tol = tol
+    cdef floating norm_min_subgrad = 0
+    cdef floating max_min_subgrad
+    cdef unsigned int active_set_ii, active_set_jj
+    cdef unsigned int ii, jj
+    cdef int n_iter = 0
+    cdef unsigned int f_iter
+    cdef UINT32_t rand_r_state_seed = rng.randint(0, RAND_R_MAX)
+    cdef UINT32_t* rand_r_state = &rand_r_state_seed
+
+    # Equivalent to Q[active_set_ii] in the diag_fisher=False version
+    cdef np.ndarray[np.float64_t, ndim=1] Q_active_set_ii = np.empty(n_active_features, dtype=np.float64)
```

Comment:
Can you use memoryviews instead of the old numpy interface? The rest of the code uses memoryviews so this would also make it more consistent.

Reply:
On hold